crates/bin/prove_block/src/main.rs

```diff
@@ -321,146 +387,113 @@ async fn main() -> Result<(), Box<dyn Error>> {
-                        panic!("legacy class (TODO)");
-                    }
-                }
-            } else {
-                log::warn!("No class available for contract {}", contract_address);
+            let generic_cc = generic_sierra_cc.compile()?;
+
+            log::debug!("============================");
+            log::debug!("class_hash (from RPC): {:x?}", class_hash);
+            log::debug!("class_hash (computed): {:x?}", generic_sierra_cc.class_hash()?);
+
+            // TODO: it seems that this ends up computing the wrong class hash...
+            let (contract_class_hash, compiled_contract_hash) = write_class_facts(
+                generic_sierra_cc,
+                generic_cc.clone(),
+                &mut initial_state.ffc.clone_with_different_hash::<PoseidonHash>(),
+            )
+            .await?;
+
+            class_hash_to_compiled_class_hash.insert(class_hash, compiled_contract_hash.into());
+            compiled_classes.insert(compiled_contract_hash.into(), generic_cc.clone());
+
+            log::debug!(
+                "Computed class_hash => compiled_class_hash: {:x?} => {:x?}",
+                contract_class_hash,
+                compiled_contract_hash
+            );
+
+            // let sierra_cc_hash = generic_sierra_cc.class_hash()?;
+            let sierra_cc_hash = generic_cc.class_hash()?;
+            log::debug!("sierra_cc_hash: {:x?}", sierra_cc_hash);
+
+            let class_hash_as_snos_hash: starknet_os_types::hash::Hash = class_hash.into();
+            if contract_class_hash != class_hash_as_snos_hash {
+                log::warn!("FFC computed different hash ({:x?}) than expected ({:?})", contract_class_hash, class_hash);
             }
         } else {
             log::warn!("No class hash available for contract {}", contract_address);
         };
     }
 
+    // query storage proofs for each accessed contract
+    let class_hashes: Vec<&Felt252> = class_hash_to_compiled_class_hash.keys().collect();
+    let class_proofs = get_class_proofs(&pathfinder_client, &args.rpc_provider, block_number, &class_hashes[..])
+        .await
+        .expect("Failed to fetch class proofs");
+
+    // write facts from class proof
+    for proof in class_proofs.values() {
+        for node in &proof.class_proof {
+            match node {
+                TrieNode::Binary { left, right } => {
+                    // log::info!("writing binary node...");
+                    let fact = BinaryNodeFact::new((*left).into(), (*right).into())?;
+                    fact.set_fact(&mut initial_state.ffc_for_class_hash).await?;
+                }
+                TrieNode::Edge { child, path } => {
+                    // log::info!("writing edge node...");
+                    let fact = EdgeNodeFact::new((*child).into(), NodePath(path.value.to_biguint()), Length(path.len))?;
+                    fact.set_fact(&mut initial_state.ffc_for_class_hash).await?;
+                }
```

Comment:
Here we dont need the previous ones (as with contract proofs)?

Not necessary to add now, but im curious :) 

If we need, lests make sure to add this in the ticket that refers to class data stuff

Reply:
By `previous` are you referring to using `block_number` instead of `block_number - 1`?

That's a good catch, and probably something really subtle since the class isn't likely to change from block to block...

Reply:
Changed to block-1 here: https://github.com/Moonsong-Labs/snos/pull/19/commits/3386814ebe0b2624604a8e7c8a5eb68a1164ab76

That said, I wonder if we actually need both like we do with the storage proofs... @odesenfans WDYT?

Reply:
Yup! But in the blocks that decalre stuff can happen! Maybe leave a comment or TODO just for us to be aware of this

Reply:
Oh, good point, that's a simple case where this would be relevant.

Reply:
https://github.com/Moonsong-Labs/snos/pull/19/commits/7271c799dbcbd32552e631e709f9901c86c610fa